logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/GetNotificationsUseCase.kt

```diff
@@ -0,0 +1,80 @@
+import com.wire.kalium.logic.data.notification.LocalNotificationMessage
+import com.wire.kalium.logic.data.notification.LocalNotificationMessageAuthor
+import com.wire.kalium.logic.data.publicuser.model.OtherUser
+import com.wire.kalium.logic.data.user.UserId
+import com.wire.kalium.logic.data.user.UserRepository
+import com.wire.kalium.logic.functional.flatMapFromIterable
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flatMapMerge
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.take
+
+interface GetNotificationsUseCase {
+    suspend operator fun invoke(): Flow<List<LocalNotificationConversation>>
+}
+
+class GetNotificationsUseCaseImpl(
+    private val messageRepository: MessageRepository,
+    private val userRepository: UserRepository,
+    private val conversationRepository: ConversationRepository
+) : GetNotificationsUseCase {
+
+    override suspend operator fun invoke(): Flow<List<LocalNotificationConversation>> {
+        return conversationRepository.getConversationsForNotifications()
+            .flatMapMerge { conversations ->
+                conversations.flatMapFromIterable { conversation ->
+                    messageRepository.getMessagesByConversationAndDate(conversation.id, conversation.lastNotificationDate ?: "")
+                        .map { messages -> ConversationWithMessages(messages, conversation) }
+                }
+            }
+            .map { it.filter { conversationWithMessages -> conversationWithMessages.messages.isNotEmpty() } }
+            .flatMapMerge { conversationsWithMessages ->
+                val authorIds = mutableSetOf<UserId>()
+
+                conversationsWithMessages.forEach { conversationAndMessages ->
+                    conversationAndMessages.messages.forEach { authorIds.add(it.senderUserId) }
+                }
+
+                authorIds.flatMapFromIterable { userId -> userRepository.getKnownUser(userId) }
+                    .map { authors ->
+                        conversationsWithMessages.map { conversationWithMessages ->
+                            val conversationId = conversationWithMessages.conversation.id
+                            val conversationName = conversationWithMessages.conversation.name ?: ""
+                            val messages = conversationWithMessages.messages.map { it.toDbNotificationMessage(authors) }
+                            val isOneToOneConversation = conversationWithMessages.conversation.type == Conversation.Type.ONE_ON_ONE
+
+                            LocalNotificationConversation(
+                                id = conversationId,
+                                name = conversationName,
+                                messages = messages,
+                                isOneToOneConversation = isOneToOneConversation
+                            )
+                        }
+                    }
+            }
+    }
```

Comment:
Can we add some inline documentation to understand better what is each step/operation doing? Otherwise it takes a bit to understand the internal logic of the use case

Reply:
good point, done